modules/generator/registry/counter.go

```diff
@@ -121,12 +137,26 @@ func (c *counter) collectMetrics(appender storage.Appender, timeMs int64, extern
 	}
 
 	for _, s := range c.series {
+		t := time.UnixMilli(timeMs)
 		// set series-specific labels
 		for i, name := range c.labels {
 			lb.Set(name, s.labelValues[i])
 		}
 
-		_, err = appender.Append(0, lb.Labels(nil), timeMs, s.value.Load())
+		// If we are about to call Append for the first time on a series, we need
+		// to first insert a 0 value to allow Prometheus to start from a non-null
+		// value.
+		if s.isNew() {
+			_, err = appender.Append(0, lb.Labels(nil), timeMs, 0)
+			if err != nil {
+				return
+			}
+			// Increment timeMs to ensure that the next value is not at the same time.
+			t = t.Add(insertOffsetDuration)
```

Comment:
I'm wondering if we could just do `timeMs - 1000`. Instead of delaying our sample, write a zero sample before it. This might trigger out-of-order writes though... I'm not sure if Prometheus checks for out-of-order writes within the whole batch or just for the serie.

Reply:
I wondered about this also, but avoided going negative offset because I wasn't sure about the interplay with Prometheus and was somewhat hard to test.  Do you think this will impact the usage having a delay on our sample?

Reply:
It adds latency between when the span arrived and when we report it has arrived, so the data becomes more inaccurate. But since we already aggregate data in the collection interval (which is typically 15s) I don't think this will be significant. So I'm fine with going for the safe options and just adding a 1s offset.